Add missing support for addition and subtraction

nilshg opened the pull request:
Inspired by the discussion in #28570.

Tests pass locally with this change. 

There are no additional tests for `missing` support in this PR, it seems to me the methods are so simple that I'm not sure what would be tested here (although I have a feeling saying things like this is why I'm not a software developer...)

nilshg committed:
Add missing support for addition and subtraction

Inspired by the discussion in #28570

nalimilan commented on stdlib/Dates/src/arithmetic.jl:
```diff
@@ -84,6 +84,12 @@ end
 (-)(x::Time, y::TimePeriod) = return Time(Nanosecond(value(x) - tons(y)))
 (+)(y::Period, x::TimeType) = x + y
 
+# Missing support
+(+)(x::T, y::Missing) where {T<:AbstractTime} = missing
+(+)(x::Missing, y::T) where {T<:AbstractTime} = missing
+(-)(x::T, y::Missing) where {T<:AbstractTime} = missing
+(-)(x::Missing, y::T) where {T<:AbstractTime} = missing
```
Even simpler:
```suggestion
# Missing support
(+)(x::AbstractTime, y::Missing) = missing
(+)(x::Missing, y::AbstractTime) = missing
(-)(x::AbstractTime, y::Missing) = missing
(-)(x::Missing, y::AbstractTime) = missing
```

pdeffebach commented:
You need `<`, `>`, `<=`, and `>=` as well, right?

nilshg committed:
Update stdlib/Dates/src/arithmetic.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

nilshg commented:
> 
> 
> You need `<`, `>`, `<=`, and `>=` as well, right?

Those are covered in `missings.jl` in Base as per discussion on Slack, right?

pdeffebach commented:
Yes, no need to add them.

nilshg committed:
Add tests for missing arithmetic

These test the exported subtypes of `AbstractTime`

nalimilan committed:
Remove trailing whitespace

nilshg committed:
Hardcode `DatePeriod` and `TimePeriod` subtypes

Avoids the usage of `subtypes` function, which would need to be imported from `InteractiveUtils` otherwise

nalimilan commented on stdlib/Dates/test/arithmetic.jl:
```diff
@@ -500,7 +500,7 @@ end
 
 @testset "Missing arithmetic" begin
-    for t ∈ [Date; Time; subtypes(DatePeriod); subtypes(TimePeriod)]
+    for t ∈ [Date Time Day Month Week Year Hour Microsecond Millisecond Minute Nanosecond Second]
```
It is more natural to use a vector rather than a matrix even if the result is the same:
```suggestion
    for t ∈ [Date, Time, Day, Month, Week, Year, Hour, Microsecond, Millisecond, Minute, Nanosecond, Second]
```

nalimilan reviewed (approved):
Thanks, looks good to me.

As discussed at #28570, I think this is justified as all types that support `+` and `-` should propagate `missing`. This is different from adding new `missing`-propagating *functions*, which I agree is problematic as it's not clear where to stop. Here the limit is very clear: add *methods* to all functions that already propagate missing values to ensure consistency.

Fun fact: this missing method was also [spotted](https://github.com/pandas-dev/pandas/issues/28095#issuecomment-537743228) by pandas developers, who have recently worked on improving their missing values handling and took Julia as a reference.

nilshg committed:
Update stdlib/Dates/test/arithmetic.jl

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

nilshg commented:
> 
> 
> Thanks, looks good to me.
> 

Thanks for your help!
